### PR TITLE
MAKE-1008: special events collector

### PR DESCRIPTION
**Pull request by tychoish**

### CODE REVIEW

File: events/collector.go [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+
+	c.current.Add(in)
+	return c.Collector.Add(c.current.MarshalDocument())
+}
+
+type noopCollector struct {
+	ftdc.Collector
+}
+
+// NewNoopCollector constructs a collector that does not sum
```
Comment (Kimchelly) [inline events/collector.go]: Isn't it kind of confusing to call it a no-op if it doesn't actually no-op?

File: events/collector.go [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+
+func (c *samplingCollector) Add(interface{}) error { return nil }
+func (c *samplingCollector) AddEvent(in *Performance) error {
+	if in == nil {
+		return errors.New("cannot add nil performance event")
+	}
+
+	if c.current == nil {
+		c.current = in
+		return c.Collector.Add(c.current.MarshalDocument())
```
Comment (Kimchelly) [inline events/collector.go]: Should this modify count before returning?

File: events/collector.go [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+func (c *randSamplingCollector) isPercent() bool {
+	if c.percent > 100 {
+		return true
+	}
+
+	if c.percent <= 0 {
+		return false
+	}
+
+	return rand.Intn(101) > (100 - c.percent)
```
Comment (Kimchelly) [inline events/collector.go]: Maybe rewrite as `rand.Intn(101) < c.percent` for clarity?

File: events/collector.go [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+func (c *samplingCollector) Add(interface{}) error { return nil }
+func (c *samplingCollector) AddEvent(in *Performance) error {
+	if in == nil {
+		return errors.New("cannot add nil performance event")
+	}
+
+	if c.current == nil {
+		c.current = in
+		c.count++
+		return c.Collector.Add(c.current.MarshalDocument())
```
Comment (julianedwards) [inline events/collector.go]: Why are we adding to the collector here? Shouldn't we check `shouldCollect` first? Or we always want to add the first performance event?

File: events/collector.go [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+
+func (c *randSamplingCollector) Add(interface{}) error { return nil }
+func (c *randSamplingCollector) AddEvent(in *Performance) error {
+	if in == nil {
+		return errors.New("cannot add nil performance event")
+	}
+
+	if c.current == nil {
+		c.current = in
+		return c.Collector.Add(c.current.MarshalDocument())
```
Comment (julianedwards) [inline events/collector.go]: again, do we want to always add the first one?

File: events/collector.go [hunk 6]
```diff
@@ -0,0 +1,255 @@
...
+	}
+
+	c.current.Add(in)
+	if c.isPercent() {
+		return c.Collector.Add(c.current.MarshalDocument())
+	}
+	return nil
+}
+
+func (c *randSamplingCollector) isPercent() bool {
```
Comment (julianedwards) [inline events/collector.go]: nit: maybe name this `shouldCollect`

File: events/collector.go (hunk 4 above)
Comment (tychoish) [reply events/collector.go]: because presumably this is the first point, and we have to have something saved in the collector to add/cumulative to.

**Pull request merged**